doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
+   The notation `x = y` is equivalent to `eq x y`.  The notation `x = y :> A`
+   gives the type of x and y explicitly.
 
-   ``forall (x``:sub:`1` ``:A``:sub:`1` ``) ... (x``:sub:`n` ``:A``:sub:`n` ``), eq term``:sub:`1` ``term``:sub:`2` ``.``
+- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
+  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
+  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
+  or :gdef:`quotient`
+  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
+  with setoids rather than constructing quotients, for which there is no specific support.
 
-   where :g:`eq` is the Leibniz equality or a registered setoid equality.
+- :gdef:`Definitional equality <definitional equality>` is equality based on the conversion
+  rules, which Coq can determine automatically.  (See :ref:`Conversion-rules`)
+  When two terms are definitionally equal, Coq knows it can
+  replace one with the other, such as with :tacn:`change` `X with Y`, among many
+  other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
+  two terms are definitionally equal.
 
-   Then :n:`rewrite @term` finds the first subterm matching `term`\ :sub:`1` in the goal,
-   resulting in instances `term`:sub:`1`' and `term`:sub:`2`' and then
-   replaces every occurrence of `term`:subscript:`1`' by `term`:subscript:`2`'.
-   Hence, some of the variables :g:`x`\ :sub:`i` are solved by unification,
-   and some of the types :g:`A`\ :sub:`1`:g:`, ..., A`\ :sub:`n` become new
-   subgoals.
+  .. todo PR: A and B are convertible iff they are definitionally equal??
 
-   .. exn:: The @term provided does not end with an equation.
-      :undocumented:
+.. _rewritingexpressions:
 
-   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
-      :undocumented:
+Rewriting with Leibniz and setoid equality
+------------------------------------------
 
-   .. tacv:: rewrite -> @term
+.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-      Is equivalent to :n:`rewrite @term`
+   .. insertprodn oriented_rewriter one_term_with_bindings
 
-   .. tacv:: rewrite <- @term
+   .. prodn::
+      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
+      one_term_with_bindings ::= {? > } @one_term {? with @bindings }
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+   .. todo PR: When I search for "rewrite (" in the .v files, I see tons of examples that I don't
+      understand.  Need to say something about them.
 
-   .. tacv:: rewrite @term in @goal_occurrences
+   Rewrites terms based on equalities.  :n:`@one_term` can be the name of a hypothesis that has
+   the form :n:`EQ @term__1 @term__2` or a lemma that has the form:
```

Comment:
I would avoid making a distinction between a hypothesis and a lemma. A lemma can be of the form `EQ term1 term2` if the terms are grounded (no free variable). This doesn't happen a lot but can exist. And a hypothesis can have binders (`forall ...`) just as a lemma.

Reply:
I think I like it this way:

![image](https://user-images.githubusercontent.com/1253341/101727600-9f171800-3a69-11eb-80e0-343084ab45be.png)